Add steerable pyramid decomposition

andreydung opened the pull request:
Based on Steerable Pyramid transforms by E. P. Simoncelli "http://www.cns.nyu.edu/~eero/steerpyr/".

It return a list of list of images representing steerable pyramid subbands.

andreydung committed:
Add steerable pyramid, based on "http://www.cns.nyu.edu/~eero/steerpyr/" paper.

andreydung committed:
Fix small bug when reconSteerable

stefanv commented:
@andreydung Please have a look at the contributor guidelines.  Code needs to be formatted according to PEP8.  Also, the unit testing and inline documentation is inadequate as it stands.  There also isn't an example.

Sorry, I know this is a lot of work, but be assured that we all have to go through the same process when making contributions!

grlee77 commented:
@andreydung:  I know it has been awhile, but is there any interest on your end in reviving this PR?  If so, I am willing to help review and move it forward. 

A user over at PyWavelets also recently inquired about this transform:
https://groups.google.com/forum/#!topic/pywavelets/QSqaH_6GBAw

I think this would fit better under the `transforms` module rather than in `filters`.

I see that various names such as `buildSCFpyr` match those of the Matlab implementation upon which this was presumably based.  The original Matlab implementation is [MIT licensed](https://github.com/LabForComputationalVision/matlabPyrTools/blob/master/LICENSE), so I think all is well in terms of borrowing from that implementation, but function and variable names should be modified to be more Pythonic (e.g. function names should not be in CamelCase as per the PEP8 guidelines mentioned above).

andreydung commented:
Yes, I'd love to have the code merged. Besides names, should I add test cases or anything else?

soupault commented:
@andreydung see https://github.com/scikit-image/scikit-image/blob/master/.github/PULL_REQUEST_TEMPLATE.md#checklist : tests, gallery examples, and PEP8 fixes. Everything else is up to you :) . Also, please, select a checkbox on the right "Allow access for maintainers" - we'll help you to make a rebase, and finalize the PR where needed.

grlee77 commented:
All public functions / classes need documentation strings in the NumPy style:
https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt

At a minimum, a short description and the `Parameters` and `Returns` sections should be present.  You can look at other functions in the library for examples of how these are typically formatted. 

For PEP8 stuff, it is easiest to set up an editor that can auto-highlight PEP8 violations for you (e.g. Spyder, SublimeText, etc.).

grlee77 commented on skimage/filters/steerable.py:
```diff
@@ -0,0 +1,210 @@
+from .._shared.utils import assert_nD
+
+def buildSteerable(image, height = 5):
+    assert (len(image.shape) == 2)
```
Do not use `assert` statements in this manner.  Test as
```Python
if len(image.shape) != 2:
    raise ValueError(...) 
```

grlee77 commented on skimage/filters/steerable.py:
```diff
@@ -0,0 +1,210 @@
+def buildSteerable(image, height = 5):
+    assert (len(image.shape) == 2)
+
+    if sampling:
```
`sampling` does not appear to be defined anywhere

grlee77 commented on skimage/filters/steerable.py:
```diff
@@ -0,0 +1,210 @@
+    if sampling:
+        s = Steerable(height)
+    else:
+        s = SteerableNoSub(height)
```
`SteerableNoSub` is not defined

andreydung committed:
move from filters to transform

andreydung committed:
hold steerable in transform

pep8speaks commented:
Hello @andreydung! Thanks for updating the PR.



 - In the file [`skimage/transform/steerable.py`](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py), following are the PEP8 issues :

> [Line 6:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L6): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 6:33](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L6): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 6:35](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L6): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 15:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L15): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 24:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L24): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 25:30](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L25): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 25:32](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L25): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 48:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L48): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 63:18](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L63): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 66:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L66): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 74:58](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L74): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 76:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L76): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (113 > 79 characters)
> [Line 79:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L79): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 79 characters)
> [Line 84:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L84): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 85:48](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L85): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 85:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L85): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 79 characters)
> [Line 91:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L91): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 92:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L92): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 106:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L106): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 119:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L119): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 120:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L120): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 124:58](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L124): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 126:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L126): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (113 > 79 characters)
> [Line 132:69](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L132): [E225](https://duckduckgo.com/?q=pep8%20E225) missing whitespace around operator
> [Line 132:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L132): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 134:62](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L134): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 134:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L134): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (103 > 79 characters)
> [Line 136:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L136): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 138:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L138): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 139:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L139): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 139:85](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L139): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 140:52](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L140): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 150:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L150): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 182:5](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L182): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 183:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L183): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 184:78](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L184): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 184:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L184): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 184:84](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L184): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 184:86](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L184): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 185:78](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L185): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 185:80](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L185): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (88 > 79 characters)
> [Line 185:84](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L185): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 185:86](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L185): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 210:41](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/steerable.py#L210): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

 - In the file [`skimage/transform/tests/test_steerable.py`](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/tests/test_steerable.py), following are the PEP8 issues :

> [Line 6:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/tests/test_steerable.py#L6): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 8:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/tests/test_steerable.py#L8): [E101](https://duckduckgo.com/?q=pep8%20E101) indentation contains mixed spaces and tabs
> [Line 8:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/tests/test_steerable.py#L8): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs
> [Line 9:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/tests/test_steerable.py#L9): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs
> [Line 10:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/tests/test_steerable.py#L10): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs
> [Line 11:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/tests/test_steerable.py#L11): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs
> [Line 12:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/tests/test_steerable.py#L12): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs
> [Line 14:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/tests/test_steerable.py#L14): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs
> [Line 14:41](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/tests/test_steerable.py#L14): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 14:43](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/tests/test_steerable.py#L14): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 16:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/tests/test_steerable.py#L16): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 17:1](https://github.com/scikit-image/scikit-image/blob/b6291d1dd350c4e3cbf4290bf7e4ac4d1e08c98f/skimage/transform/tests/test_steerable.py#L17): [W191](https://duckduckgo.com/?q=pep8%20W191) indentation contains tabs

andreydung committed:
updated doc strings

andreydung commented:
@grlee77 I updated with docstrings, would still need more work on testing and providing doc examples.

How could I run the specific tests for steerable only? For now I'm running `make test` and it takes a long time.

grlee77 commented:
> How could I run the specific tests for steerable only? 

You should be able to run `py.test` on specific folders or files:
` py.test skimage/transform/tests/test_steerable.py` to run only tests from that file
`py.test skimage/transforms` would find all tests in the `transforms` module.

andreydung committed:
fix dtype error

andreydung committed:
added plot examples

andreydung commented:
@grlee77 I finished the code based on your reviews (fixed format, added docstrings, added multiple test cases and one example in doc folder). Could you please look again at the latest commit? Let me know if anything else needed to be done.

When writing this version, I tried to have the exact result as Matlab version (unit tested all functions between Python and Matlab version). Thus efficiency was not paid much attention, for example this could have been written iteratively instead of recursively. Also the reconstruction is not always exact, it depends on the shape of the image (ideally should be power of two).

On a side note, how do I run the examples code in doc/examples folder? It keeps importing the official version, not the local one.

andreydung committed:
fix typos in example

andreydung committed:
fixed all PEP8 issues

andreydung committed:
resolve conflict of __init__.py

andreydung committed:
remove carve

grlee77 commented:
Thanks.  I will review this again when I get a chance.

> On a side note, how do I run the examples code in doc/examples folder? It keeps importing the official version, not the local one. 

I am not sure how you have currently set things up for development, but I would recommend using an environment specifically dedicated to `scikit-image` development.  Two common ways to do this are via either [virtualenv](https://virtualenv.pypa.io/en/stable/) or using `conda` environments in [Anaconda](http://anaconda.org).  In this dedicated environment, I do not install a realease version of scikit-image at all.  Instead, I do an in-place development build (e.g. via `pip install -e .` from within the source folder of `scikit-image`).  This way there will be no conflicting "official version" and "local version".

details on using conda for this here:  
https://uoa-eresearch.github.io/eresearch-cookbook/recipe/2014/11/20/conda/
https://www.continuum.io/blog/developer-blog/python-packages-and-environments-conda

details on using virtualenv for this here:  
http://python-guide-pt-br.readthedocs.io/en/latest/dev/virtualenvs/

andreydung committed:
Merge branch 'master' into steerable

andreydung committed:
fixed steerable example

andreydung committed:
Merge branch 'steerable' of https://github.com/andreydung/scikit-image into steerable

andreydung committed:
move into folder

andreydung committed:
fix PEP8 issue

andreydung committed:
fix title overline too short

andreydung commented:
@grlee77 I rebased, and code passed. Now the problem is docstring of the example that Travis CI does not pass. 

How do I check this locally to save time? Run make html in the folder?

andreydung committed:
fix sphinx error

andreydung committed:
fix PEP8 and sphinx

andreydung commented:
@grlee77 All checks have passed. Could you please have a look at the PR again?

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,339 @@
+        height : height of the Steerable Decomposition
+        (including high pass and low pass)
+        """
+        self.nbands = 4
```
`nbands` should be an optional parameter to __init__ so the user can set it as desired.  It can then also be an optional parameter in `build_steerable` and `recon_steerable`.

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,339 @@
+    Parameters
+    ----------
+    coeff : subbands of Steerable decomposition,
+            stored as a list of 'height' sublists,
```
only indent 4 spaces relative to `coeff`

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,339 @@
+            nresdft = self.recon_pyr_level(
+                coeff[1:], nlog_rad, Xrcos, Yrcos, nangle)
+
+            res = np.fft.fftshift(np.fft.fft2(nresdft))
```
`res` is computed but not used?  can this line be removed?

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,339 @@
+    ctrm = np.ceil((m + 0.5) / 2).astype(int)
+    ctrn = np.ceil((n + 0.5) / 2).astype(int)
+    xv, yv = np.meshgrid((np.array(range(n)) + 1 - ctrn) / (n / 2),
+                         (np.array(range(m)) + 1 - ctrm) / (m / 2))
```
add `sparse=True` as an argument to `mesh grid` to save memory

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,339 @@
+    """
+    Given a 1D curve defined by X, Y, convert it to a 2D mask
+    """
+    out = np.interp(mask.flatten(), X, Y)
```
use `ravel` rather than `flatten` to avoid a copy

grlee77 commented on skimage/transform/tests/test_steerable.py:
```diff
@@ -0,0 +1,48 @@
+    out = steerable.recon_steerable(coeff)
+
+    assert_array_almost_equal(img_as_float(im), out, decimal=1)
+
```
The tests above look reasonable.  It is probably also worth adding some other dtypes to at least one of these tests as it seems only `uint8` inputs are tested.  floats in the range [0, 1] and a binary image perhaps?

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,339 @@
+    return X, Y
+
+
+def point_op(mask, Y, X):
```
I would rename this `_point_op` as it will not be part of the public API

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,339 @@
+    return log_rad, angle
+
+
+def rcos_curve(width, position):
```
I would rename this `_rcos_curve` as it will not be part of the public API

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,339 @@
+        return np.fft.ifft2(np.fft.ifftshift(outdft)).real
+
+
+def base(m, n):
```
I would rename this `_base` as it will not be part of the public API

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,339 @@
+
+    Parameters
+    ----------
+    m, n : size of the desired grid
```
Please check the doctoring format here and in a couple of the other utility functions.  These type should come after the colon with the description indented on the following line(s).  e.g.
```
m, n : int
    size of the desired grid
```

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,339 @@
+
+        im = img_as_float(im)
+
+        M, N = im.shape
```
`r, c` or `nr, nc` to indicate rows/columns rather than `M, N` for better consistency with skim age conventions

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,339 @@
+                raise ValueError(
+                    "Size of intermediate sublist should be %d" % self.nbands)
+
+        M, N = coeff[0].shape
```
`r, c` or `nr, nc` to indicate rows/columns rather than `M, N` for better consistency with skim age conventions

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,339 @@
+        """
+        self.nbands = 4
+        self.height = height
+        self.isSample = True
```
I haven't yet looked what `isSample` does, but should it also be an optional input to `__init__` so the user can set it as desired?

andreydung replied:
Steerable pyramid has two versions: with sampling and without sampling. For now I decided to include sampling version only. isSampling is an attribute to decide what version it is.

No sampling version could be added easily by inheriting from Steerable class. Shall I add it then?

grlee77 replied:
Only if there is some advantage to the without sampling version.  If not, then I would just leave it as is.

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,339 @@
+        Xrcos, Yrcos = rcos_curve(1, -0.5)
+        Yrcos = np.sqrt(Yrcos)
+        YIrcos = np.sqrt(1 - Yrcos * Yrcos)
+
```
can swap order of lines 86-87 to avoid `Yrcos*Yrcos`:
```python
YIrcos = np.sqrt(1 - Yrcos)
Yrcos = np.sqrt(Yrcos)
```

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,339 @@
+        lo0mask = point_op(log_rad, YIrcos, Xrcos)
+        hi0mask = point_op(log_rad, Yrcos, Xrcos)
+
+        imdft = np.fft.fftshift(np.fft.fft2(im))
```
Perhaps all `fft2` and `ifft2` calls should use the `norm` argument via a class parameter such as `self.fft_norm`.  By default the numpy FFTs use `norm=None`, but `norm='oath'` might be desirable to have energy in the coefficients that is similar to that of the input image.

andreydung replied:
Adding norm='ortho' break the test cases (I added this to all fft2 and ifft2 functions).

grlee77 reviewed (changes requested):
I did a partial review.  I still have to look into the internals of the Steerable class a bit more closely, but overall this is much improved and seems to be operating as expected.

I think it should also be reasonably easy to extend this to work for `multichannel` by specifying `axes=(0, 1)` on the FFTs and so forth., but you don't have to add that as part of this PR.

andreydung committed:
fix docs and add test cases for floats

andreydung committed:
fix steerable.py without breaking tests

andreydung committed:
all PEP8 are fixed

andreydung committed:
stricter tests

andreydung committed:
try to fix sphinx

andreydung committed:
fix 2 lines in example for PEP8

andreydung committed:
add steerable to init

andreydung commented:
@grlee77 I updated PR based on your reviews and added lots of test cases. 

Appveyor check failed though, and I don't really know how to fix this. Could you please help me out?

andreydung committed:
add sparse to meshgrid

andreydung commented:
@grlee77 Could you please have a look at the PR again? Let me know if anything else should be added.

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -228,28 +257,33 @@ def recon_pyr_level(self, coeff, log_rad, Xrcos, Yrcos, angle):
 
     def recon_scf_pyramid(self, coeff):
-        """
+        """ Reconstruc the image from its Steerable decomposition
```
typo: missing a t on Reconstruct.  Also, please add a period to the end of the sentence.

grlee77 commented on skimage/transform/tests/test_steerable.py:
```diff
@@ -0,0 +1,94 @@
+    coeff = steerable.build_steerable(im)
+    out = steerable.recon_steerable(coeff)
+
+    assert_array_almost_equal(img_as_float(im), out, decimal=2)
```
Is this relatively low number of decimals of agreement to be expected for this transform?  Have you done any validation vs. results from the Matlab implementation itself?

andreydung replied:
Yes, all functions are unit tested between Matlab and Python.

grlee77 replied:
What is in the tests included here just seems to verify primarily that roundtrip decomposition/reconstruction is correct.
When you say "all functions are unit tested vs. Matlab", do you mean that you did this offline as an independent verification?  It is possible to archive expected outputs from Matlab to use during testing.  We have some tests that do that sort of thing in PyWavelets (see, [here](https://github.com/PyWavelets/pywt/blob/master/pywt/tests/test_matlab_compatibility.py) for example), but I don't know that we need to go to that level of trouble here.  Visually the coefficients in the example look as one would expect and the round-trip seems okay, so I think that is probably enough.

andreydung committed:
fixed docstrings, increase decimals in testcases

andreydung commented:
@grlee77 Updated tests with stricter agreement, except for asymmetric cases which can only match with 1 decimal of accuracy.

Also I'm not sure why this PR fails with Python 3.4.

grlee77 commented:
The one CI failure doesn't look related to this PR.  I just restarted it to see if it was just a fluke.

grlee77 commented:
> Also the reconstruction is not always exact, it depends on the shape of the image (ideally should be power of two).

Is it really power of two or is it any integer multiple of `2**height`?  (i.e. power of two is only necessary in the case where the maximum possible number of levels of decomposition are performed).

There is currently no check that the user provided `height` is reasonable.  For example, if I change round-trip examples with input of size 128x128 to use an excessive `height=10`, no error is raised, but there is a large error in the round-trip result.  I think we need to raise a `ValueError` if the specified `height` exceeds what is possible for the input array.

If accuracy is likely to be poor for some shapes, perhaps we should raise a warning for the user in those instances?  The user would then know that they may benefit from padding up to an appropriate shape via `np.pad` prior to performing the decomposition.

andreydung committed:
add checking of the height

andreydung commented:
@grlee77 I added the checking of height. 

It turns out that previously, the second dimension of asymmetric shape (113,**29**) is less than 2**height (where height is 5) thus the reconstruction error is high. Now I reduced the height to 3, and get 5 decimal matching of the reconstruction!

andreydung committed:
remove test for warning

andreydung committed:
remove warn import

andreydung commented:
@grlee77 any updates on this? Thanks!

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -84,6 +84,13 @@ def __init__(self, height=5, nbands=4):
             number of orientations in Steerable decomposition
         """
+
+        if not (isinstance(height, int) and isinstance(nbands, int)):
```
isinstance with `int` is too restrictive  (`isinstance(np.int32(5), int)` returns False on most machines)

`isinstance(height, np.integer)` would match all integer types, but I tend to favor the following form as it allows floats with integer value as well:
`(height % 1 == 0)`

If you go the later route, you probably want to set, `self.height = int(height)` below so integer-like floats get converted to actual integers.

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -97,8 +104,12 @@ def build_scf_pyramid(self, im):
+
+        if min(M, N) < 2**self.height:
+            raise ValueError("Image is not big enough to "
+                             "apply Steerable decomposition")
```
I would expand this message to include some helpful specifics such as:

`"For a  {} level decomposition, all image dimensions must be at least size {}.".format(height, 2**height)"`

grlee77 commented:
Probably should not have hit the "approve changes" button above just yet as there were still a couple of minor comments pending (see the last two above).  I thought perhaps that applied to my original review only, but it seems it has now marked the two recent comments as "outdated" as well.

grlee77 commented:
> any updates on this? Thanks!

At this point the PR needs to await a second reviewer before it can be merged.  It will have to be rebased again at the end to clean up a couple things in the history (merge master into branch, etc), but I can do that for you when it is time to merge.

grlee77 commented:
I don't know what is going on with the one Travis failure.  From the message it seems related to the intial `pip` call for installing dependencies (not sure why astropy is showing up in the error message?):
https://travis-ci.org/scikit-image/scikit-image/jobs/238978316#L5827

While looking at that log, I also noticed a couple of unused imports in the example are picked up by flake8:
https://travis-ci.org/scikit-image/scikit-image/jobs/238978316#L5714-L5715

andreydung committed:
fix minor checks, rewrite example to be clearer, catch a few bugs

andreydung commented:
@grlee77: I updated based on your comments.

I also changed the structure of coeff: previously first and last element of coeff is numpy array, and now I moved them inside a list. This makes the structure more uniform: all elements are list, and if flattening of the structure could be done easier.

I also rewrote the steerable example to be cleaner.

andreydung committed:
int conversion in setting

andreydung commented:
@soupault Could you please have a quick look at the PR? Let me know if anything else is needed.

andreydung commented:
@grlee77 Hi Gregory, is there any update on this?

soupault commented:
@andreydung I've put this on my short list. Will review once I get a free timeslot. :)

grlee77 commented:
I may be somewhat less available until the 28th.  I never did a close comparison to the original publication, but the unit tests here at least verify that round trip decomposition/reconstruction is working and I believe @andreydung stated he had also verified individual functions vs. the original author's Matlab implementations.  

Thanks for sticking with it @andreydung.  The process of getting things merged can take a little while, but we will get there eventually!

andreydung commented:
@soupault I'm looking forward to your review!

andreydung commented:
@grlee77 @soupault Hi all, just a reminder for the reviews!

stefanv commented on doc/examples/transform/plot_steerable.py:
```diff
@@ -0,0 +1,87 @@
+from skimage.transform import steerable
+
+
+def normalize(im):
```
Use `http://scikit-image.org/docs/dev/api/skimage.exposure.html#skimage.exposure.rescale_intensity`

stefanv commented on doc/examples/transform/plot_steerable.py:
```diff
@@ -0,0 +1,87 @@
+    return (im - im.min()) / (im.max() - im.min())
+
+
+def visualize(coeff):
```
No docstring

stefanv commented on doc/examples/transform/plot_steerable.py:
```diff
@@ -0,0 +1,87 @@
+
+def visualize(coeff):
+    rows, cols = coeff[1][0].shape
+    Norients = len(coeff[1])
```
Python naming convention is nr_orients or num_orients or, at worst, n_orients

stefanv commented on doc/examples/transform/plot_steerable.py:
```diff
@@ -0,0 +1,87 @@
+    return out
+
+
+# create an image of a disk
```
Use disk function from morphology?

stefanv commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,406 @@
+
+def build_steerable(image, height=5, nbands=4):
+    """ Construct a Steerable subband decomposition of a gray scale image.
+
```
A second, detailed paragraph is needed, describing the algorithm, what it is used for, etc.  Implementation details should be added under Notes.  This also holds for the function below, `recon_steerable`, or at least a See Also added to `build_steerable`.

stefanv commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,406 @@
+    Parameters
+    ----------
+    image : 2-D array
+        input image
```
Input image.

stefanv commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,406 @@
+    image : 2-D array
+        input image
+    height : integer, optional
+        Height of the steerable decomposition. This includes the counting of
```
This description is vague.

stefanv commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,406 @@
+    height : integer, optional
+        Height of the steerable decomposition. This includes the counting of
+        low pass and high pass subbands.
+    nbands : integer, optional
```
n_bands, nr_bands, or just bands

stefanv commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,406 @@
+    Returns
+    -------
+    coeff : list of lists of numpy array
+        subbands of Steerable decomposition,
```
Formatting, capitalization, wrapping.

stefanv commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,406 @@
+
+
+class Steerable:
+    """Steerable Pyramid: a translation and rotation invariant free wavelet.
```
Details lacking.

stefanv commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,406 @@
+        """
+        Parameters
+        ----------
+        height : int
```
Same comment as earlier.

stefanv commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,406 @@
+        self.height = int(height)
+
+    def build_scf_pyramid(self, im):
+        """
```
Single line description missing.

stefanv commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,406 @@
+        Parameters
+        ----------
+        im : 2-D array
+            input gray scale image
```
Capitalization.

stefanv commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,406 @@
+        return coeff
+
+    def _build_pyr_level(self, lodft, log_rad, angle, Xrcos, Yrcos, ht):
+        """
```
Single line description

stefanv commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,406 @@
+
+        return coeff
+
+    def _recon_pyr_level(self, coeff, log_rad, Xrcos, Yrcos, angle):
```
It looks like these methods share repeated code segments; is it possible to refactor out some common functionality?

andreydung replied:
I agree that they're somewhat duplicate, but refactoring will cleanup a few lines only (since there's difference between build and recon here and there). Also the common code does not stand out as a separate function, i.e taking them out will make reading them more difficult since there are no more context.

jni replied:
I agree with @stefanv. Despite the difference, there are many opportunities for simplification. For example, this code is identical between the two methods:

```python
           Xrcos = Xrcos - 1
           himask = _point_op(log_rad, Yrcos, Xrcos)
           lutsize = 1024
           Xcosn = np.pi * \
               np.array(range(-(2 * lutsize + 1), (lutsize + 2))) / lutsize
           order = self.n_bands - 1
           const = np.power(2, 2 * order) * \
               np.square(np.math.factorial(order)) / \
               (self.n_bands * np.math.factorial(2 * order=)
```

That is only the most obvious example. I think you should definitely look at improving the code sharing between the two. It will massively reduce the maintainability burden in the future.

stefanv commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,406 @@
+
+def _base(m, n):
+    """
+    Helper function that create a grid containing radius from center and angle.
```
`_base` is uninformative; Description here is not detailed enough.

stefanv reviewed (changes requested):
Thank you for your contribution, @andreydung!  I appreciate the care you've taken to write clean and functional code.

I have done a fairly pedantic review of code style and structure, so please excuse the numerous comments.  The requested comments may seem extraneous, but I can assure you that those coming after us who have to maintain the code will thank us :)

grlee77 commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,406 @@
+        "The Steerable Pyramid: A Flexible Architecture
+        for Multi-Scale Derivative Computation."
+        http://www.cns.nyu.edu/~eero/steerpyr/
+    """
```
please expand this reference here (and at top of the module) to include the publication details:
```
    .. [1] Simoncelli, E.P. & Freeman, W.T.
       (1995). The Steerable Pyramid: A Flexible Architecture for Multi-Scale
       Derivative Computation. In Proc. 2nd IEEE International Conf. on Image
       Proc., vol.III pp. 444-447, Oct 1995. 
       http://www.cns.nyu.edu/~eero/steerpyr/,
       http://www.cns.nyu.edu/pub/eero/simoncelli95b.pdf,
       DOI:10.1109/ICIP.1995.537667
```

andreydung replied:
Done!

andreydung committed:
detailed docstring, use skimage functions to generate samples

andreydung commented:
@grlee77 @stefanv sorry for the long delay, I updated code based on your reviews. Let me know if there's anything else needed. Let's get this PR done soon!

stefanv commented:
I don't have time to review this this week; @scikit-image/core any takers?

andreydung committed:
fixing missing self

andreydung commented:
@scikit-image/core I'm looking forward to the reviews! Thanks for the time!

jni commented on doc/examples/transform/plot_steerable.py:
```diff
@@ -0,0 +1,96 @@
+       Proc., vol.III pp. 444-447, Oct 1995. 
+       http://www.cns.nyu.edu/~eero/steerpyr/,
+       http://www.cns.nyu.edu/pub/eero/simoncelli95b.pdf,
+       DOI:10.1109/ICIP.1995.537667
```
Could you add a short explanation of the use of these pyramids, other than plotting the coefficients?

jni commented on doc/examples/transform/plot_steerable.py:
```diff
@@ -0,0 +1,96 @@
+from skimage.morphology import disk
+
+
+def visualize(coeff):
```
This might be worth putting into the `steerable` module, or class.

jni replied:
(Possibly as `visualize_steerable_pyramid_coefficients` or something similarly descriptive but more concise.)

jni commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,425 @@
+        self.n_bands = int(n_bands)
+        self.height = int(height)
+
+    def build_scf_pyramid(self, im):
```
What does "scf" stand for?

jni commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,425 @@
+
+    """
+    N = 256
+    X = np.pi * np.array(range(-N - 1, 2)) / 2 / N
```
PEP8: please use lowercase variable names **unless** referring to matrices (in the linear algebra sense).

jni commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,425 @@
+    return log_rad, angle
+
+
+def _rcos_curve(width, position):
```
Rename to `_raised_cosine_curve`

jni commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,425 @@
+    return X, Y
+
+
+def _point_op(mask, Y, X):
```
I have no idea what this function is doing. Please consider renaming it and providing an example in the docstring?

jni commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,425 @@
+        YIrcos = np.sqrt(1 - Yrcos)
+        Yrcos = np.sqrt(Yrcos)
+
+        lo0mask = _point_op(log_rad, YIrcos, Xrcos)
```
Why is the variable named `lo0mask`?

jni commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,425 @@
+
+            coeff = self._build_pyr_level(
+                lodft, log_rad, angle, Xrcos, Yrcos, ht - 1)
+            coeff.insert(0, orients)
```
Inserting into a Python list is much more expensive than appending. I suggest you build the pyramid in reverse order, and then at the top level, return `list(reversed(coeff))`

jni commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,425 @@
+            Helper matrices to help create the DFT mask
+        Xrcos, Yrcos: 2-D array
+            Represents the desired filter
+        ht: int
```
Rename to `height`.

jni commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,425 @@
+            # ================== Subsample lowpass ============================
+            dims = np.array(lodft.shape)
+
+            lostart = np.ceil((dims + 0.5) / 2) - \
```
PEP8: put parentheses around the whole expression rather than using `\`

jni commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,425 @@
+            lostart = lostart.astype(np.int32)
+            loend = loend.astype(np.int32)
+
+            log_rad = log_rad[lostart[0]:loend[0], lostart[1]:loend[1]]
```
minor: 
```python
box = (slice(lostart[0], loend[0]), slice(lostart[1], loend[1]))
log_rad = log_rad[box]
angle = angle[box]
lodft = lodft[box]
```

jni commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,425 @@
+            Subbands of Steerable decomposition,
+            stored as a list of 'height' sublists,
+            Sublists correspond to decreasing radius level in Steerable pyramid
+            The first sublist contains the high pass subband.
```
First sublist or first element of a sublist?

jni replied:
Sometimes for this kind of explanation I find it helpful to use sample indices: `coeff[i][j] contains the subband of orientation j at height i.`, for example.

jni commented on skimage/transform/steerable.py:
```diff
@@ -0,0 +1,425 @@
+        ----------
+        coeff : list of lists of numpy array
+            Subbands of Steerable decomposition,
+            stored as a list of 'height' sublists,
```
`,` -> `.`

jni reviewed (changes requested):
Hi @andreydung! Thanks for the PR and for the ping on the mailing list. ;) I've left a few comments, but my impression of what this PR needs is:

- At least one more round of refactoring to ensure less code repetition. Hiding code repetition behind good function names can dramatically simplify understanding of the code (which I'm still fuzzy about). For example, `_logradius_angle_grid` is a great name and helps me understand what's happening wherever it is called.
- another round of refactoring improving some variable names. For example, `ht` should be changed to `height`, and `lo0mask` should be changed to (maybe?) `low_circle_mask`?
- The gallery example should include some description about how the steerable pyramid can be used to achieve certain goals.

I'm also not sure that you need the `SteerablePyramid` class, since it does not seem to hold on to any useful data for reuse? If so, consider removing it and moving all the code into the corresponding function.

grlee77 commented:
@andreydung : Is there still interest in finishing up this PR? I think it is pretty close (just some additional comments above from @jni)

Also, FYI, I recently discovered another related implemenation:
It is an ITK module which implements steerable wavelets in n-dimensions (under an Apache 2.0 license). It appears to have Python wrappers, but I haven't tried to use them yet.
https://github.com/InsightSoftwareConsortium/ITKIsotropicWavelets

andreydung commented:
@grlee77 I will push the PR by this weekend. I'll have a look at the ITK implementations as well.

billbrod commented:
Is there still interest in integrating the steerable pyramid into scikit-image? I've been part of [two](https://github.com/LabForComputationalVision/pyrtools/blob/main/pyrtools/pyramids/SteerablePyramidFreq.py) separate python [implementations](https://github.com/LabForComputationalVision/plenoptic/blob/main/plenoptic/simulate/canonical_computations/steerable_pyramid_freq.py) of the steerable pyramid (one using numpy and the C code underlying the original matlab implementation, one using pytorch), and would be interested in helping getting a version included in scikit-image if folks still think it's useful. I've also tried putting [together](https://github.com/billbrod/steerable-pyramid-tutorial) some [different](https://pyrtools.readthedocs.io/en/latest/tutorials/03_steerable_pyramids.html) tutorial [materials](https://plenoptic.readthedocs.io/en/latest/tutorials/models/03_Steerable_Pyramid.html), which I'd be happy to adapt if they'd be useful as well.

stefanv commented:
Sorry for letting this languish; the team is quite overloaded.

I appreciate those resources, @billbrod, that puts the feature into better context. I see that the original paper is from 1995. This makes me think that this work may have been superseded by wavelets? Even if not, I wonder if it is better scoped for PyWavelets.

The translation and rotation invariant features are neat, and a gallery example that makes the feature click (like those in Z Püspöki et al., "Transforms and operators for directional bioimage analysis: a survey") would be useful.

I'm pushing back gently, but with a clear application (I couldn't find that in your notes), and a demonstrated need out there, we'd certainly be interested in incorporating it into the library.

billbrod commented:
No worries, I understand how that goes. 

My sense is that the steerable pyramid shares characteristics with both wavelets (found in pywavelets) and Gaussian + Laplacian pyramids (found in scikit-image). I think of it as more similar to the pyramids in how it's used (and it feels a natural progression: Gaussian (lowpass, unoriented) -> Laplacian (bandpass, unoriented) -> Steerable (bandpass, oriented)), so my inclination was that it belonged in scikit-image, but I'd defer to you all. What's the difference in scope between the two packages?

I come from Eero Simoncelli's lab, so I have a bias towards the steerable pyramid :grin:. For me, coming from vision science / neuroscience, it's useful when building a representation like that found in the primate primary visual cortex (V1, as schematized in figure 7 of [scholarpedia's V1 article](http://www.scholarpedia.org/article/Area_V1)), for which I want to easily get the outputs of oriented, band-pass filters across multiple scales, across the whole image, and I want filters in quadrature pair. So that's the application for which I've used it, but I recognize that doesn't mean it necessarily belongs in scikit-image.

I agree about a nice demonstration of translation and rotation invariance. To be clear, you're talking about more than what's shown [here](https://plenoptic.readthedocs.io/en/latest/tutorials/models/03_Steerable_Pyramid.html#Putting-the-%22Steer%22-in-Steerable-Pyramid), right? More like what's discussed [here](https://github.com/LabForComputationalVision/plenoptic/issues/204)?

I'm not very familiar with wavelets, but steerable pyramids are (really) over-complete (whereas most wavelet are not?) and pretty much always used in the multi-scale context (whereas wavelets are often used at a single scale?). I believe this over-completeness leads to a smoother representation of orientation, and thus is less likely to alias. I can try to put together a more detailed comparison of steerable pyramids and wavelets if that would help determine whether steerable pyramids belong in scikit image?

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
